DataFormats/TrackReco/interface/TrackExtraBase.h

```diff
@@ -25,6 +25,18 @@ class TrackExtraBase
     using TrajParams = std::vector<LocalTrajectoryParameters>;
     using Chi2sFive = std::vector<unsigned char>;
 
+    /// class which implemebts begin() and end() to enable range-based loop on RecHits
+    class TrackingRecHitCollectionWindow {

```

Comment:
I think `TrackingRecHitRange` would be a better name (closer to standard naming convention). In which case it would probably be better to be defined in `DataFormats/TrackingRecHit`. (we could even consider a new class template that could then be instantiated in `DataFormats/TrackingRechit/interface/TrackingRecHitFwd.h`)

Reply:
Indeed, that name is shorter and better. So I understand correctly that you'd define the class in `DataFormats/TrackingRechit/interface/TrackingRecHitFwd.h`? But where would you put the template? Actually I don't know if that pattern reoccurs in CMSSW and the template in necessary. Thanks for your advice!

Reply:
> So I understand correctly that you'd define the class in `DataFormats/TrackingRechit/interface/TrackingRecHitFwd.h`

That file is for forward declarations so I wouldn't define the `TrackingRecHitRange` class there (unless it would be a template instantiation).

> But where would you put the template?

I'm thinking `edm::ConstRange<T>` (with `T` being the iterator type) in either `FWCore/Utilities` or `DataFormats/Common`. @Dr15Jones, would you have any input?

> Actually I don't know if that pattern reoccurs in CMSSW and the template in necessary.

I don't recall seeing this pattern, but many times in the past I would have wanted to use range-for for iterator pairs obtained with something else that `begin()`+`end()` pair (which reminds me that I may have introduced some ad-hoc range classes somewhere, but can't remember where). I just believe this pattern would have potential to be used elsewhere as well.

Reply:
Ok! So that means if it wouldn't be a template it belongs in `DataFormats/TrackingRechit/interface/TrackingRecHit.h` without the Fwd? Hope I understand this well. Alright, let's see what Chris says.

I just recalled there would indeed many use-cases for such a class. There is for example not only the `*Begin()/*End()` pattern that was seen in the tracks, but sometimes people just created pairs to imitate a range, like here:

https://github.com/cms-sw/cmssw/pull/25591/files#diff-2718c8ec8e06c5b5679d465a9c81b6e8R426

If I'm not mistaken I have seen such things elsewhere.

Reply:
Since the location of the file is not so important after all, I just suggested how the `edm::ConstRange<T>` could look like.

Reply:
> So that means if it wouldn't be a template it belongs in `DataFormats/TrackingRechit/interface/TrackingRecHit.h` without the Fwd? 

Or in a separate header.

> I just recalled there would maybe be more use-cases for such a template. There is for example not only the *Begin()/*End() pattern that was seen in the tracks, but sometimes people just created pairs to imitate a range, like here:

Yeah, there are plenty of those. Many (if not all) of those predate C++11 and range-for.